win,msi: change InstallScope to perMachine

joaocgreis opened the pull request:
Currently, installing with the MSI always asks for Administrator permissions, as mentioned in #7629 . Explicitly installing per machine only adds the benefit that shortcuts are placed in ProgramData thus being accessible to all users and also solving #5849 .

The second commit moves the installation path in the registry to Local Machine. The registry stores HKLM in different places for 32 and 64 bit applications, so the installer will not suggest the old path when upgrading from 32 to 64 bit version, solving #5592 and #25087 .

/cc @piscisaureus

joaocgreis committed:
win,msi: change InstallScope to perMachine

The MSI install scope was set to the WiX default, which is per-user.
However, with UAC, it could not be installed by a standard user because
InstallPrivileges is elevated by default, hence the install scope
should be set to per-machine. Furthermore, the default install path is
a per-machine location and setting the system path requires
administrator privileges.

By changing the InstallScope to perMachine, Start Menu shortcuts are
placed in ProgramData and not the installing user's AppData folder,
making the shortcuts available to other users. This also fixes the
installation when AppData is a network folder.

The custom action is necessary to allow upgrades. Since a per-machine
MSI cannot upgrade an application installed per-user, the custom action
checks if there is going to be an upgrade to a previous version
installed per-user and sets the installation as per-user to allow
upgrading. Hence, the advantages of installing per-machine will only
apply in fresh installations.

Fixes #5849
Fixes #7629

joaocgreis committed:
win,msi: correct installation path registry keys

Since install is per machine only, installation path should be stored
in local machine instead of current user. The registry stores HKLM in
different places for 32 and 64 bit applications, so the installer will
not suggest the old path when upgrading from 32 to 64 bit version.

Fixes #5592
Fixes #25087

orangemocha commented:
LGTM, but would still be good if someone else (@piscisaureus ?) could take a look.

piscisaureus commented:
The problem I see with this is the following:

In order for NPM to work, `%appdata%\npm` needs to be added to the path.
That path is user-specific.
Setting the installation scope to global would add the wrong path.

joaocgreis commented:
@piscisaureus Thanks for reviewing. In all my tests, the `%appdata%\npm` still gets added to the user path, but only for the user that installs. This is the same behavior that happens without these changes. Do you have a specific situation in mind in which the wrong path is added?

orangemocha commented:
> In order for NPM to work, %appdata%\npm needs to be added to the path.

@piscisaureus , thanks for reviewing! As @joaocgreis said, %appdata%\npm still gets added to the user path as before. OK to land this?

piscisaureus commented:
> @piscisaureus , thanks for reviewing! As @joaocgreis said, %appdata%\npm still gets added to the user path as before. OK to land this?

If that's true, then yes.

orangemocha commented:
Triple-checked that it does. Will land shortly...

orangemocha committed:
test: mark test-signal-unregister as flaky

This test just failed on Ubuntu in Jenkins, for a change that
is 100% Windows-specific.

orangemocha commented:
FYI, adding a commit to this PR mark test-signal-unregister as flaky. It failed on Ubuntu and that's obviously unrelated to this change:
https://jenkins-iojs.nodesource.com/job/node-test-commit-linux/28/nodes=ubuntu1504-64/tapTestReport/test.tap-576/

orangemocha commented:
Actually, I will mark that test as flaky in a separate PR.

orangemocha commented:
Landed in 14db629497d0eb97a5f0fbb70cf3e534e460deec.

orangemocha commented:
This should probably go in the converged repo. @jasnell what do we need to do to make it so?
